{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### task description\n",
    "brain eQTL enrichment analysis\n",
    "\n",
    "*  First dataset is this: http://eqtl.brainseq.org/phase2/\n",
    "`SupplementaryTable15_eQTL.tar.gz` in hg38 with Gencode v25\n",
    "Tables with the significant eQTL associations (FDR < 1%) for DLPFC, HIPPO and the brain region interaction (DLPFC vs HIPPO) at the gene, exon, exon-exon junction and transcript expression levels.  \n",
    "\n",
    "    * snp: SNP ID.\n",
    "    * feature_id: expression feature ID. You might also want the EnsemblGeneID column (Ensembl gene ID) or the Symbol one (gene symbol, when available).\n",
    "    * statistic: eQTL t-statistic computed by MatrixEQTL.\n",
    "    * pvalue: p-value.\n",
    "    * FDR: FDR adjusted p-value.\n",
    "    * beta: eQTL beta coefficient.\n",
    "    \n",
    "    * SNP annotation file: `wget https://libd-brainseq2.s3.us-east-2.amazonaws.com/BrainSeqPhaseII_snp_annotation.txt.gz`\n",
    "\n",
    "* Second dataset from [Common Minds Consortium]( https://www.nimhgenetics.org/resources/commonmind): https://www.sugarsync.com/pf/D7756315_09685381_064295\n",
    "`CMC_MSSM-Penn-Pitt_DLPFC_mRNA_eQTL-adjustedSVA.txt` in hg19\n",
    "\n",
    "\n",
    "    \n",
    "* Random SNP sets generated by Kevin Luo: `CN_20_ASoC_FDR0.05_matchedSNP` and `NSC_20_ASoC_FDR0.05_matchedSNP`, in hg19\n",
    "\n",
    "    * The file named “SNPmatch_100_matchedSNPs.txt.gz” has a table of input SNPs and matched SNPs.\n",
    "    The first column is the input SNPs, the second is the number of SNPs in the SNPsnap database satisfied our thresholds, the following 100 columns are 100 matching SNPs sampled from the SNPsnap database that satisfied our thresholds.\n",
    "    * The file named “snpsnap_input_exist.txt” saved the information for the SNPs that exist in the database.\n",
    "    * The file named “unincluded_snps.txt” show the SNPs that are not exist in the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### prepare input file to snpsnap (generate random set of SNPs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/project2/xinhe/simingz/neuron_atac_seq/HiC\n"
     ]
    }
   ],
   "source": [
    "cd /home/simingz/neuron_atac_seq/HiC/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $1\"\\t\"$3}' CN_20_ASoC_FDR0.05.bed > ../eQTL/CN_20_ASoC_FDR0.05.bed.chrpos.hg38\n",
    "!awk '{print $1\"\\t\"$3}' NSC_20_ASoC_FDR0.05.bed > ../eQTL/NSC_20_ASoC_FDR0.05.bed.chrpos.hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/project2/xinhe/simingz/neuron_atac_seq/eQTL\n"
     ]
    }
   ],
   "source": [
    "cd ../eQTL/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyliftover import LiftOver\n",
    "lo = LiftOver('hg38', 'hg19')\n",
    "def lofile(f, outf, lo):\n",
    "    with open(f) as f1, open(outf, 'w') as outf1:\n",
    "        for line in f1.readlines():\n",
    "            chrom = line.split('\\t')[0]\n",
    "            pos = int(line.strip().split('\\t')[1])\n",
    "            out = lo.convert_coordinate(chrom, pos)\n",
    "            if len(out) != 0:\n",
    "                outf1.write(out[0][0].split('r')[1] + ':' + str(out[0][1]) + '\\n')\n",
    "\n",
    "def lofile2(f, outf, lo):\n",
    "    with open(f) as f1, open(outf, 'w') as outf1:\n",
    "        for line in f1.readlines():\n",
    "            chrom = line.split('\\t')[0]\n",
    "            end = int(line.strip().split('\\t')[2]) # different from lofile above\n",
    "            newline = line.split('\\t')[3: len(line.split('\\t'))]\n",
    "            out = lo.convert_coordinate(chrom, end)\n",
    "            if len(out) != 0:\n",
    "                outf1.write('\\t'.join([out[0][0],str(out[0][1] -1), str(out[0][1])]+ newline))\n",
    "\n",
    "\n",
    "lofile('CN_20_ASoC_FDR0.05.bed.chrpos.hg38', 'CN_20_ASoC_FDR0.05.bed.chrpos.hg19', lo)\n",
    "lofile('NSC_20_ASoC_FDR0.05.bed.chrpos.hg38', 'NSC_20_ASoC_FDR0.05.bed.chrpos.hg19', lo)\n",
    "lofile2('/home/simingz/neuron_atac_seq/HiC/CN_20_ASoC_FDR0.05.bed', '/home/simingz/neuron_atac_seq/HiC/CN_20_ASoC_FDR0.05.bed.hg19', lo)\n",
    "lofile2('/home/simingz/neuron_atac_seq/HiC/NSC_20_ASoC_FDR0.05.bed', '/home/simingz/neuron_atac_seq/HiC/NSC_20_ASoC_FDR0.05.bed.hg19', lo)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### prepare Brain eQTL 1\n",
    "ChrX is ignored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "eQTLf1 = 'BrainSeqPhaseII_eQTL_FDR1perc_DLPFC_gene.txt'\n",
    "annof = 'BrainSeqPhaseII_snp_annotation.txt'\n",
    "annodict = {}\n",
    "with open(annof, 'r') as anno:\n",
    "    csv_reader = csv.reader(anno, delimiter = '\\t')\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "            annodict.update({row[0]:row[4] + ':' + row[5]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl1 = []\n",
    "fail = []\n",
    "with open(eQTLf1, 'r') as f1:\n",
    "    csv_reader = csv.reader(f1, delimiter = '\\t')\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        if len(row[0].split(\":\")) > 1 and row[0].split(\":\")[0] != 'chrX':\n",
    "            try:\n",
    "                int(row[0].split(':')[0])\n",
    "                eqtl1.append('chr' + row[0].split(':')[0] + ':' + row[0].split(':')[1])\n",
    "            except ValueError:\n",
    "                try:\n",
    "                    eqtl1.append(annodict[row[0]])\n",
    "                except KeyError:\n",
    "                    fail.append(row[0])\n",
    "                \n",
    "eqtl1 = set(eqtl1)\n",
    "fail= set(fail)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of eQTL: 824893\n",
      "Number failed to include: 0\n"
     ]
    }
   ],
   "source": [
    "print('Number of eQTL: ' + str(len(eqtl1)) + '\\nNumber failed to include: ' + str(len(fail)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### prepare brain eQTL2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "eQTLf2 = 'CMC_MSSM-Penn-Pitt_DLPFC_mRNA_eQTL-adjustedSVA.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl2 = []\n",
    "with open(eQTLf2, 'r') as f2:\n",
    "    csv_reader = csv.reader(f2, delimiter = ' ')\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        try:\n",
    "            eqtl2.append(row[7])\n",
    "        except IndexError:\n",
    "            print(row)\n",
    "\n",
    "eqtl2 = set(eqtl2)                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of eQTL: 2070780\n"
     ]
    }
   ],
   "source": [
    "print('Number of eQTL: ' + str(len(eqtl2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### target SNP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "snpf1 = 'CN_20_ASoC_FDR0.05.bed.chrpos.hg19'\n",
    "snp1 = []\n",
    "with open(snpf1, 'r') as f1:\n",
    "    csv_reader = csv.reader(f1)\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        snp1.append('chr' + row[0])\n",
    "snp1 = set(snp1)      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of SNPs: 5610\n"
     ]
    }
   ],
   "source": [
    "print('Number of SNPs: ' + str(len(snp1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### target SNP2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "snpf2 = 'NSC_20_ASoC_FDR0.05.bed.chrpos.hg19'\n",
    "snp2 = []\n",
    "with open(snpf2, 'r') as f2:\n",
    "    csv_reader = csv.reader(f2)\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        snp2.append('chr' + row[0])\n",
    "snp2 = set(snp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of SNPs: 3545\n"
     ]
    }
   ],
   "source": [
    "print('Number of SNPs: ' + str(len(snp2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP1 matched random SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "snprf1 = 'CN_20_ASoC_FDR0.05_matchedSNP/SNPmatch_100_matchedSNPs.txt'\n",
    "snpr1 = []\n",
    "with open(snprf1, 'r') as f1:\n",
    "    csv_reader = csv.reader(f1, delimiter=\"\\t\")\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        snpr1.extend(['chr' + i for i in row[2:102]])\n",
    "snpr1 = set(snpr1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of SNPs: 422013\n"
     ]
    }
   ],
   "source": [
    "print('Number of SNPs: ' + str(len(snpr1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP2 matched random SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "snprf2 = 'NSC_20_ASoC_FDR0.05_matchedSNP/SNPmatch_100_matchedSNPs.txt'\n",
    "snpr2 = []\n",
    "with open(snprf2, 'r') as f2:\n",
    "    csv_reader = csv.reader(f2, delimiter=\"\\t\")\n",
    "    next(csv_reader)\n",
    "    for row in csv_reader:\n",
    "        snpr2.extend(['chr' + i for i in row[2:102]])\n",
    "snpr2 = set(snpr2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of SNPs: 280230\n"
     ]
    }
   ],
   "source": [
    "print('Number of SNPs: ' + str(len(snpr2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For SNP set1, enrichment in eQTL dataset1 results:\n",
      "Percent of SNP is eQTL in SNP set1: 0.14367201426024956\n",
      "Percent of SNP is eQTL in matched random set: 0.1066933957010803\n",
      "odds ratio: 1.4047370927845446\n",
      "p value: 1.6877869556860068e-17\n"
     ]
    }
   ],
   "source": [
    "mystat = [[len(snp1 & eqtl1), len(snp1 - eqtl1)], [len(snpr1 & eqtl1), len(snpr1 - eqtl1)]]\n",
    "odds, pvalue = stats.fisher_exact(mystat)\n",
    "print('For SNP set1, enrichment in eQTL dataset1 results:')\n",
    "print('Percent of SNP is eQTL in SNP set1: ' + str(len(snp1 & eqtl1)/len(snp1)))\n",
    "print('Percent of SNP is eQTL in matched random set: ' + str(len(snpr1 & eqtl1)/len(snpr1)))\n",
    "print('odds ratio: ' + str(odds) + \"\\np value: \"+ str(pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For SNP set2, enrichment in eQTL dataset1 results:\n",
      "Percent of SNP is eQTL in SNP set2: 0.12609308885754583\n",
      "Percent of SNP is eQTL in matched random set: 0.10493523177390002\n",
      "odds ratio: 1.2307199661101298\n",
      "p value: 7.190754513783032e-05\n"
     ]
    }
   ],
   "source": [
    "mystat = [[len(snp2 & eqtl1), len(snp2 - eqtl1)], [len(snpr2 & eqtl1), len(snpr2 - eqtl1)]]\n",
    "odds, pvalue = stats.fisher_exact(mystat)\n",
    "print('For SNP set2, enrichment in eQTL dataset1 results:')\n",
    "print('Percent of SNP is eQTL in SNP set2: ' + str(len(snp2 & eqtl1)/len(snp2)))\n",
    "print('Percent of SNP is eQTL in matched random set: ' + str(len(snpr2 & eqtl1)/len(snpr2)))\n",
    "print('odds ratio: ' + str(odds) + \"\\np value: \"+ str(pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For SNP set1, enrichment in eQTL dataset2 results:\n",
      "Percent of SNP is eQTL in SNP set1: 0.3762923351158645\n",
      "Percent of SNP is eQTL in matched random set: 0.28609308244058834\n",
      "odds ratio: 1.5054922495116099\n",
      "p value: 2.33319980128235e-47\n"
     ]
    }
   ],
   "source": [
    "mystat = [[len(snp1 & eqtl2), len(snp1 - eqtl2)], [len(snpr1 & eqtl2), len(snpr1 - eqtl2)]]\n",
    "odds, pvalue = stats.fisher_exact(mystat)\n",
    "print('For SNP set1, enrichment in eQTL dataset2 results:')\n",
    "print('Percent of SNP is eQTL in SNP set1: ' + str(len(snp1 & eqtl2)/len(snp1)))\n",
    "print('Percent of SNP is eQTL in matched random set: ' + str(len(snpr1 & eqtl2)/len(snpr1)))\n",
    "print('odds ratio: ' + str(odds) + \"\\np value: \"+ str(pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For SNP set2, enrichment in eQTL dataset2 results:\n",
      "Percent of SNP is eQTL in SNP set2: 0.34809590973201693\n",
      "Percent of SNP is eQTL in matched random set: 0.280391107304714\n",
      "odds ratio: 1.370400473687971\n",
      "p value: 2.7802319096376094e-18\n"
     ]
    }
   ],
   "source": [
    "mystat = [[len(snp2 & eqtl2), len(snp2 - eqtl2)], [len(snpr2 & eqtl2), len(snpr2 - eqtl2)]]\n",
    "odds, pvalue = stats.fisher_exact(mystat)\n",
    "print('For SNP set2, enrichment in eQTL dataset2 results:')\n",
    "print('Percent of SNP is eQTL in SNP set2: ' + str(len(snp2 & eqtl2)/len(snp2)))\n",
    "print('Percent of SNP is eQTL in matched random set: ' + str(len(snpr2 & eqtl2)/len(snpr2)))\n",
    "print('odds ratio: ' + str(odds) + \"\\np value: \"+ str(pvalue))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:testotherpy3]",
   "language": "python",
   "name": "conda-env-testotherpy3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
